cvat/apps/dataset_manager/tracks_counter.py

```diff
@@ -0,0 +1,200 @@
+# Copyright (C) CVAT.ai Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from collections import defaultdict
+from typing import Callable
+
+from django.db import transaction
+
+from cvat.apps.engine.models import Job, LabeledTrack, ShapeType
+from cvat.apps.engine.utils import defaultdict_to_regular
+
+from .util import linear_sort_shapes
+
+
+class TracksCounter:
+    """
+    The class implements counting of VISIBLE shapes in tracks.
+
+    It keeps also track type and label id in internal data structure.
+    These values are not used directly inside this class.
+    However this class is inherited in our private code and this is small trade-off to reduce code duplication.
+    """
+
+    def __init__(self):
+        self._tracks_per_job = {}
+        self._stop_frames_per_job = {}
+
+    def count_track_shapes(self, job_id: int, track_id: int):
+        track = self._tracks_per_job.get(job_id, {}).get(track_id)
+        manual = 0
+        interpolated = 0
+
+        if track is not None and track["shapes"]:
+            prev_frame = track["shapes"][0]["frame"]
+            prev_is_outside = track["shapes"][0]["outside"]
+
+            if not prev_is_outside:
+                manual += 1
+
+            for shape in track["shapes"][1:]:
+                frame = shape["frame"]
+                is_outside = shape["outside"]
+                assert frame > prev_frame
+
+                if not prev_is_outside:
+                    # -1 means that current keyframe is not interpolated frame
+                    interpolated += frame - prev_frame - 1
+                if not is_outside:
+                    manual += 1
+
+                prev_is_outside = is_outside
+                prev_frame = frame
+
+            if not prev_is_outside and prev_frame < self._stop_frames_per_job[job_id]:
+                interpolated += self._stop_frames_per_job[job_id] - prev_frame
+
+        return {
+            "manual": manual,
+            "interpolated": interpolated,
+        }
+
+    def load_tracks_from_db(
+        self,
+        parent_labeledtrack_qs_filter: Callable,
+        child_labeledtrack_qs_filter: Callable,
+    ):
+        with transaction.atomic():
```

Comment:
The transaction doesn't seem to be doing anything here.

Reply:
Transaction guarantees us fetched parent tracks and child tracks are relevant to each other

Reply:
Are you sure about such guarantees if no locks is taken? AFAIK, transactions by themselves only guarantee atomicity, which means automatic rollback in the case of an error.

Reply:
Ok, documentation says specifically about delete/update operations only. 
Removed this block